web/interceptor/access_control_test.go

```diff
@@ -0,0 +1,417 @@
+
+	tm, err := auth.NewTokenManager(db, "test")
+	if err != nil {
+		t.Fatal(err)
+	}
+
+	lis := bufconn.Listen(BufSize)
+	bufDialer := func(context.Context, string) (net.Conn, error) {
+		return lis.Dial()
+	}
+	opt := grpc.ChainUnaryInterceptor(
+		interceptor.UnaryUserVerifier(logger, tm),
+		interceptor.AccessControl(logger, tm),
+	)
+	s := grpc.NewServer(opt) // skipcq: GO-S0902
+	qf.RegisterQuickFeedServiceServer(s, ags)
+
+	go func() {
+		if err := s.Serve(lis); err != nil {
+			t.Errorf("Server exited with error: %v", err)
+			return
+		}
+	}()
+
+	ctx := context.Background()
+	conn, err := grpc.DialContext(ctx, "bufnet", grpc.WithContextDialer(bufDialer), grpc.WithTransportCredentials(insecure.NewCredentials()))
+	if err != nil {
+		t.Fatalf("Failed to dial bufnet: %v", err)
+	}
+	defer conn.Close()
+
+	client := qf.NewQuickFeedServiceClient(conn)
+
+	courseAdmin := qtest.CreateAdminUser(t, db, "fake")
+	groupStudent := qtest.CreateNamedUser(t, db, 2, "group student")
+	student := qtest.CreateNamedUser(t, db, 3, "student")
+	user := qtest.CreateNamedUser(t, db, 4, "user")
+	admin := qtest.CreateFakeUser(t, db, 5)
+	admin.IsAdmin = true
+	if err := db.UpdateUser(admin); err != nil {
+		t.Fatal(err)
+	}
+
+	course := &qf.Course{
+		Code:             "test101",
+		Year:             2022,
+		Provider:         "fake",
+		OrganizationID:   1,
+		OrganizationPath: "testorg",
+		CourseCreatorID:  courseAdmin.ID,
+	}
+	if err := db.CreateCourse(courseAdmin.ID, course); err != nil {
+		t.Fatal(err)
+	}
+	qtest.EnrollStudent(t, db, groupStudent, course)
+	qtest.EnrollStudent(t, db, student, course)
+	group := &qf.Group{
+		CourseID: course.ID,
+		Name:     "Test",
+		Users:    []*qf.User{groupStudent},
+	}
+	if err := db.CreateGroup(group); err != nil {
+		t.Fatal(err)
+	}
+
+	assignment := &qf.Assignment{
+		CourseID: course.ID,
+		Name:     "Test Assignment",
+		Order:    1,
+	}
+	if err := db.CreateAssignment(assignment); err != nil {
+		t.Fatal(err)
+	}
+	if err := db.CreateSubmission(&qf.Submission{
+		AssignmentID: assignment.ID,
+		UserID:       groupStudent.ID,
+	}); err != nil {
+		t.Fatal(err)
+	}
+
+	f := func(t *testing.T, id uint64) context.Context {
+		token, err := tm.NewAuthCookie(id)
+		if err != nil {
+			t.Fatal(err)
+		}
+		return qtest.WithAuthCookie(ctx, interceptor.AuthTokenString(token.Value))
+	}
+	courseAdminContext := f(t, courseAdmin.ID)
+	groupStudentContext := f(t, groupStudent.ID)
+	studentContext := f(t, student.ID)
+	userContext := f(t, user.ID)
+	adminContext := f(t, admin.ID)
+
+	freeAccessTest := accessTests{
+		{"admin", courseAdminContext, 0, course.ID, 0, true},
+		{"student", studentContext, 0, course.ID, 0, true},
+		{"student", groupStudentContext, 0, course.ID, 0, true},
+		{"user", userContext, 0, course.ID, 0, true},
+		{"non-teacher admin", adminContext, 0, course.ID, 0, true},
+	}
+	testUnrestrictedAccess(client, t, freeAccessTest)
```

Comment:
Please use subtests instead of testXYZ functions:
```go
	for _, tt := range freeAccessTest {
		t.Run("UnrestrictedAccess/"+tt.name, func(t *testing.T) {
			_, err := client.GetUser(tt.ctx, &qf.Void{})
			verifyAccess(t, err, tt.access, "GetUser", tt.name)
			_, err = client.GetCourse(tt.ctx, &qf.CourseRequest{CourseID: tt.courseID})
			verifyAccess(t, err, tt.access, "GetCourse", tt.name)
			_, err = client.GetCourses(tt.ctx, &qf.Void{})
			verifyAccess(t, err, tt.access, "GetCourses", tt.name)
		})
	}
```
This makes it much easier to identify which tests were run and also allows to run only a selected subtest when debugging, e.g.,:
```sh
% go test -v -run TestAccessControl/Unrestricted
```

PS: Since `tt.name` is now part of the subtest name, it is no longer needed by the `verifyAccess` function.